File: Python/other/Longest_Consecutive_Subsequence.py
"""
Given an array of positive integers. Find the length of the longest sub-sequence such that 
elements in the subsequence are consecutive integers,
the consecutive numbers can be in any order.

Example:
Input: {2,6,1,9,4,5,3}
Output: 6

Explanation:
The consecutive numbers here
are 1, 2, 3, 4, 5, 6. These 6 
numbers form the longest consecutive
subsquence.

"""

def LongestConsecutiveSubsequence(arr):
    #let take dict save the the element 
    dict={}
    
    for l in arr:
        if l not in dict.keys():
            dict[l]=1

    sub=[0]*len(arr)
    #check  present or not then increase the count with 1 
    for i in range(0,len(arr)):

        key=arr[i]

        while key in dict.keys():
            key=key+1

        sub[i]=key-arr[i]
    #result
    return sorted(sub,reverse=True)[0]

if __name__ == '__main__':
    arr = list(map(int,input("Enter the number: ").split()))
    print("The longest consective subsequence is ",LongestConsecutiveSubsequence(arr))

'''
Time complexity : O(n^n) 
Space complexity : O(n^n) 

INPUT:-
Enter the number: 36 41 56 35 44 33 34 92 43 32 42 

OUTPUT:-
The Longest Consecutive Subsequence is  5

'''

File: C-Plus-Plus/other/Longest_Consecutive_Subsequence.cpp
/* Find the length of the longest sub-sequence. */
#include<bits/stdc++.h>
using namespace std;

/* Function to Find the length of the longest sub-sequence. */
int solve()
{
    /* Input n = size of array */
    int n;
    cin >> n;

    int arr[n];

    /* Input values in array */
    for (int i = 0; i < n; ++i)
    {
        cin >> arr[i];
    }

    /* Input array element in set*/
    set<int> s;

    for (int i = 0; i < n; ++i)
    {
        s.insert(arr[i]);
    }

    int ans = 0;
    for (int i=0; i<n; i++) 
    { 
        /* Check the element is starting point of subsequences
           simply check arr[i]-1 in set */
        if (s.find(arr[i]-1) == s.end()) 
        { 
            /*check next elements in the sequence */
            int j = arr[i]; 
            while (s.find(j) != s.end()) 
                j++; 
  
            /* Store length in ans*/
            ans = max(ans, j - arr[i]); 
        } 
    } 
    /* Return final Length */
    return ans;


}

int main()
{
    int length = solve();

    cout << "Length of the Longest subsequence is : " << length << endl;
    return 0;
}

/*
Test cases :
    
    Input 1 :
        
        7
        1 9 3 10 4 20 2

    Output 1 :

        Length of the Longest subsequence is : 4

    Input 2 :

        11
        36 41 56 35 44 33 34 92 43 32 42
    
    Output 2 :
    
        Length of the Longest subsequence is : 5
    
    Time complexity: O(n)
    Space Complexity: O(1)
*/
